File: app/src/main/java/dev/bltucker/nanodegreecapstone/storydetail/StoryDetailFragment.java
package dev.bltucker.nanodegreecapstone.storydetail;

import android.content.Intent;
import android.graphics.Color;
import android.net.Uri;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.support.v4.graphics.drawable.DrawableCompat;
import android.support.v4.view.MenuItemCompat;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.ShareActionProvider;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Toast;

import java.util.ArrayList;

import javax.inject.Inject;

import dev.bltucker.nanodegreecapstone.R;
import dev.bltucker.nanodegreecapstone.databinding.FragmentStoryDetailBinding;
import dev.bltucker.nanodegreecapstone.injection.DaggerInjector;
import dev.bltucker.nanodegreecapstone.models.Comment;
import dev.bltucker.nanodegreecapstone.models.Story;
import dev.bltucker.nanodegreecapstone.storydetail.injection.StoryDetailFragmentModule;

public class StoryDetailFragment extends Fragment implements StoryDetailView {

    static final String STORY_BUNDLE_KEY = "story";

    private static final String DETAIL_STORY_BUNDLE_KEY = "detailStory";

    @Inject
    StoryCommentsAdapter commentsAdapter;

    @Inject
    StoryDetailViewPresenter presenter;

    @Inject
    DetailStoryProvider detailStoryProvider;

    ShareActionProvider shareActionProvider;
    MenuItem shareMenuItem;

    private DetailStory detailStory;
    private FragmentStoryDetailBinding binding;


    public StoryDetailFragment() {
        // Required empty public constructor
    }

    public static StoryDetailFragment newInstance(@Nullable Story selectedStory) {
        StoryDetailFragment storyDetailFragment = new StoryDetailFragment();
        Bundle args = new Bundle();
        args.putParcelable(STORY_BUNDLE_KEY, selectedStory);
        storyDetailFragment.setArguments(args);
        return storyDetailFragment;
    }


    @Override
    public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
        super.onCreateOptionsMenu(menu, inflater);
        inflater.inflate(R.menu.menu_detail_fragment, menu);

        shareMenuItem = menu.findItem(R.id.menu_item_share_story);
        shareMenuItem.setVisible(false);
        shareActionProvider = (ShareActionProvider) MenuItemCompat.getActionProvider(shareMenuItem);

        MenuItem saveItem = menu.findItem(R.id.menu_item_save_story);
        DrawableCompat.setTint(saveItem.getIcon(), Color.WHITE);
    }

    @Override
    public void onPrepareOptionsMenu(Menu menu) {
        super.onPrepareOptionsMenu(menu);

        if (detailStory.hasStory() && shareActionProvider != null) {
            setupShareActionProvider();
            shareMenuItem.setVisible(true);
        }
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {

        if (item.getItemId() == R.id.menu_item_save_story) {
            presenter.onSaveStoryClick(detailStory);
            return true;
        }

        return super.onOptionsItemSelected(item);
    }

    private void setupShareActionProvider() {
        Intent shareIntent = new Intent();
        shareIntent.setAction(Intent.ACTION_SEND);
        shareIntent.setType("text/plain");
        shareIntent.putExtra(Intent.EXTRA_SUBJECT, detailStory.getTitle());
        shareIntent.putExtra(Intent.EXTRA_TEXT, detailStory.getUrl());
        shareActionProvider.setShareIntent(shareIntent);
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setHasOptionsMenu(true);
        DaggerInjector.getApplicationComponent()
                .storyDetailComponent(new StoryDetailFragmentModule(this))
                .inject(this);

        initializeDetailStory(savedInstanceState);
    }

    private void initializeDetailStory(Bundle savedInstanceState) {
        //TODO HANDLE NULL STORIES FOR TABLET MODE
        if (savedInstanceState != null) {
            detailStory = savedInstanceState.getParcelable(DETAIL_STORY_BUNDLE_KEY);
        } else {
            detailStory = detailStoryProvider.getDetailStory((Story) getArguments().getParcelable(STORY_BUNDLE_KEY), new ArrayList<Comment>());
            if(detailStory.hasStory()){
                //TODO kick off comment sync?
            }
        }
    }

    @Override
    public void onSaveInstanceState(Bundle outState) {
        super.onSaveInstanceState(outState);
        outState.putParcelable(DETAIL_STORY_BUNDLE_KEY, detailStory);
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        binding = FragmentStoryDetailBinding.inflate(inflater, container, false);
        binding.headerInclude.readButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                presenter.onReadButtonClicked();
            }
        });
        return binding.getRoot();
    }

    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
        binding.commentListRecyclerview.setLayoutManager(new LinearLayoutManager(getContext()));
        commentsAdapter.setDetailStory(detailStory);
        binding.commentListRecyclerview.setAdapter(commentsAdapter);
        if (null == savedInstanceState) {
            presenter.onViewCreated(this, detailStory);
        } else {
            presenter.onViewRestored(this, detailStory);
        }
    }

    @Override
    public void onResume() {
        super.onResume();
        presenter.onViewResumed(this);
    }

    @Override
    public void onPause() {
        presenter.onViewPaused();
        super.onPause();
    }

    @Override
    public void onDestroy() {
        presenter.onViewDestroyed();
        super.onDestroy();
    }

    @Override
    public void showStory() {
        if (getActivity() != null) {
            getActivity().invalidateOptionsMenu();
        }

        if (!detailStory.hasStory()) {
            return;
        }

        binding.headerInclude.storyTitleTextview.setText(detailStory.getTitle());
        binding.headerInclude.storyUrlTextview.setText(detailStory.getUrl());
        binding.headerInclude.posterNameTextview.setText(String.format(getString(R.string.by_poster), detailStory.getPosterName()));
        binding.headerInclude.scoreTextview.setText(String.format(getString(R.string.story_score), detailStory.getScore()));
        if (null == detailStory.getUrl()) {
            binding.headerInclude.readButton.setVisibility(View.INVISIBLE);
        } else {
            binding.headerInclude.readButton.setVisibility(View.VISIBLE);
        }
    }

    @Override
    public void showStoryPostUrl() {
        if (detailStory.hasStory() && detailStory.getUrl() != null) {
            Intent browserIntent = new Intent(Intent.ACTION_VIEW, Uri.parse(detailStory.getUrl()));
            startActivity(browserIntent);
        }
    }

    @Override
    public void showStorySaveConfirmation() {
        Toast.makeText(getContext(), getString(R.string.story_saved), Toast.LENGTH_LONG).show();
    }

    @Override
    public void showEmptyView() {
        binding.commentListRecyclerview.setVisibility(View.INVISIBLE);
        binding.emptyViewContainer.setVisibility(View.VISIBLE);
        binding.headerInclude.detailHeaderCardview.setVisibility(View.INVISIBLE);
    }
}

File: app/src/main/java/dev/bltucker/nanodegreecapstone/storydetail/StoryDetailFragment.kt
package dev.bltucker.nanodegreecapstone.storydetail

import android.arch.lifecycle.ViewModelProviders
import android.content.Intent
import android.graphics.Color
import android.net.Uri
import android.os.Bundle
import android.support.design.widget.Snackbar
import android.support.v4.app.Fragment
import android.support.v4.graphics.drawable.DrawableCompat
import android.support.v4.view.MenuItemCompat
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.ShareActionProvider
import android.view.*
import android.widget.Toast
import dev.bltucker.nanodegreecapstone.R
import dev.bltucker.nanodegreecapstone.common.ApplicationViewModelsFactory
import dev.bltucker.nanodegreecapstone.databinding.FragmentStoryDetailBinding
import dev.bltucker.nanodegreecapstone.common.injection.DaggerInjector
import dev.bltucker.nanodegreecapstone.common.injection.qualifiers.UI
import dev.bltucker.nanodegreecapstone.common.models.Comment
import dev.bltucker.nanodegreecapstone.common.models.Story
import dev.bltucker.nanodegreecapstone.storydetail.injection.StoryDetailFragmentModule
import dev.bltucker.nanodegreecapstone.storydetail.injection.StoryDetailFragmentModule.Companion.DETAIL_STORY_BUNDLE_KEY
import dev.bltucker.nanodegreecapstone.storydetail.injection.StoryDetailFragmentModule.Companion.STORY_BUNDLE_KEY

import io.reactivex.Observer
import io.reactivex.Scheduler
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.disposables.Disposable
import javax.inject.Inject

class StoryDetailFragment : Fragment() {

    @Inject
    lateinit var commentsAdapter: StoryCommentsAdapter

    @Inject
    lateinit var storyCommentsSyncer: StoryCommentsSyncer

    @Inject
    lateinit var applicationViewModelsFactory: ApplicationViewModelsFactory

    @field:[Inject UI]
    lateinit var uiScheduler: Scheduler

    @Inject
    @JvmField
    var story: Story? = null

    internal var shareActionProvider: ShareActionProvider? = null

    internal lateinit var shareMenuItem: MenuItem

    private lateinit var binding: FragmentStoryDetailBinding

    private lateinit var storyDetailViewModel: StoryDetailViewModel

    private val modelSubscriptions = CompositeDisposable()

    override fun onCreateOptionsMenu(menu: Menu?, inflater: MenuInflater?) {
        super.onCreateOptionsMenu(menu, inflater)
        inflater!!.inflate(R.menu.menu_detail_fragment, menu)

        shareMenuItem = menu!!.findItem(R.id.menu_item_share_story)
        shareMenuItem.isVisible = false
        shareActionProvider = MenuItemCompat.getActionProvider(shareMenuItem) as ShareActionProvider

        val saveItem = menu.findItem(R.id.menu_item_save_story)
        DrawableCompat.setTint(saveItem.icon, Color.WHITE)
    }

    override fun onPrepareOptionsMenu(menu: Menu?) {
        super.onPrepareOptionsMenu(menu)
        if (shareActionProvider != null && story != null) {
            setupShareActionProvider()
            shareMenuItem.isVisible = true
        }
    }

    override fun onOptionsItemSelected(item: MenuItem?): Boolean {
        if (item!!.itemId == R.id.menu_item_save_story) {
            if (story != null) {
                storyDetailViewModel.onSaveStoryClick(story!!)
            }
            return true
        }

        return super.onOptionsItemSelected(item)
    }

    private fun setupShareActionProvider() {
        val shareIntent = Intent()
        shareIntent.action = Intent.ACTION_SEND
        shareIntent.type = "text/plain"
        shareIntent.putExtra(Intent.EXTRA_SUBJECT, story!!.getTitle())
        shareIntent.putExtra(Intent.EXTRA_TEXT, story!!.getUrl())
        shareActionProvider!!.setShareIntent(shareIntent)
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setHasOptionsMenu(true)
        DaggerInjector.getApplicationComponent()
                .storyDetailComponent(StoryDetailFragmentModule(this, savedInstanceState))
                .inject(this)
        storyDetailViewModel = ViewModelProviders.of(this, applicationViewModelsFactory).get(StoryDetailViewModel::class.java)
        lifecycle.addObserver(storyCommentsSyncer)
    }

    override fun onSaveInstanceState(outState: Bundle?) {
        super.onSaveInstanceState(outState)
        outState!!.putParcelable(DETAIL_STORY_BUNDLE_KEY, story)
    }

    override fun onCreateView(inflater: LayoutInflater?, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        binding = FragmentStoryDetailBinding.inflate(inflater!!, container, false)
        binding.headerInclude?.readButton?.setOnClickListener { _ -> showStoryPostUrl() }
        return binding.root
    }

    override fun onViewCreated(view: View?, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        binding.commentListRecyclerview.layoutManager = LinearLayoutManager(view!!.context)
        binding.commentListRecyclerview.adapter = commentsAdapter
    }

    override fun onStart() {
        super.onStart()

        if (story != null) {
            showStory()

            storyDetailViewModel.getObservableComments(story!!.getId())
                    .observeOn(uiScheduler)
                    .subscribe(object : Observer<Array<Comment>> {
                        override fun onSubscribe(d: Disposable) {
                            modelSubscriptions.add(d)
                        }

                        override fun onNext(comments: Array<Comment>) {
                            commentsAdapter.updateComments(comments)
                        }

                        override fun onError(e: Throwable) {
                            showGenericError("Error loading story comments")
                        }

                        override fun onComplete() {}
                    })

            storyDetailViewModel.getObservableReadLaterSuccessStatus()
                    .observeOn(uiScheduler)
                    .subscribe(object : Observer<Boolean> {
                        override fun onNext(saveSuccess: Boolean) {
                            if (saveSuccess) {
                                Toast.makeText(context, getString(R.string.story_saved), Toast.LENGTH_LONG).show()
                            }
                        }

                        override fun onSubscribe(d: Disposable) {
                            modelSubscriptions.add(d)
                        }

                        override fun onError(e: Throwable) {}

                        override fun onComplete() {}
                    })
        } else {
            showEmptyView()
        }

    }

    private fun showGenericError(errorMessage: String) {
        Snackbar.make(binding.coordinatorLayout, errorMessage, Snackbar.LENGTH_INDEFINITE).show()
    }

    override fun onStop() {
        modelSubscriptions.clear()
        super.onStop()
    }

    fun showStory() {
        if (activity != null) {
            activity.invalidateOptionsMenu()
        }

        if (null == story) {
            return
        }

        binding.headerInclude?.storyTitleTextview?.text = story!!.getTitle()
        binding.headerInclude?.storyUrlTextview?.text = story!!.getUrl()
        binding.headerInclude?.posterNameTextview?.text = String.format(getString(R.string.by_poster), story!!.getPosterName())
        binding.headerInclude?.scoreTextview?.text = String.format(getString(R.string.story_score), story!!.getScore())
        if (null == story!!.getUrl()) {
            binding.headerInclude?.readButton?.visibility = View.INVISIBLE
        } else {
            binding.headerInclude?.readButton?.visibility = View.VISIBLE
        }
    }

    fun showStoryPostUrl() {
        if (story != null && story!!.getUrl() != null) {
            val browserIntent = Intent(Intent.ACTION_VIEW, Uri.parse(story!!.getUrl()))
            startActivity(browserIntent)
        }
    }

    fun showEmptyView() {
        binding.commentListRecyclerview.visibility = View.INVISIBLE
        binding.emptyViewContainer.visibility = View.VISIBLE
        binding.appBarLayout.visibility = View.INVISIBLE
        binding.headerInclude?.detailHeaderCardview?.visibility = View.INVISIBLE
    }

    companion object {

        fun newInstance(selectedStory: Story?): StoryDetailFragment {
            val storyDetailFragment = StoryDetailFragment()
            val args = Bundle()
            args.putParcelable(STORY_BUNDLE_KEY, selectedStory)
            storyDetailFragment.arguments = args
            return storyDetailFragment
        }
    }
}
